util: Replace use of locale dependent atoi(…) with locale-independent std::from_chars(…) (C++17)

practicalswift opened the pull request:
Replace use of locale dependent `atoi(…)` with locale-independent `std::from_chars(…)` (C++17).

laanwj commented:
Note that we have a function `ParseInt32` (as well as 64 and UInt variants) for this and I once tried to use it in more places, see #17385.

laanwj commented on src/util/strencodings.h:
```diff
@@ -69,7 +71,25 @@ std::string EncodeBase32(const std::string& str, bool pad = true);
+    }
+    auto [_, error_condition] = std::from_chars(s.data(), s.data() + s.size(), result);
+    if (error_condition != std::errc{}) {
+        return 0;
```
Please, do not just return 0 in case of a parse error but use explicit error feedback. Most C++17ish would be to return an `Option`.

laanwj replied:
I think you need to check the returned pointer against `last` as well, otherwise it ignores any remainder.

practicalswift replied:
I agree on both for a sane safe interface, but since this function (`LocaleIndependentAtoi`) is meant as a quirk-by-quirk compatible to the not-entirely-sane `atoi(…)` I'm afraid we need to return a `T` (instead of `std::optional<T>`) and ignore any remainder, no? :)

laanwj commented:
>  locale-independent std::from_chars(…) (C++17).

This is good to hear at least. From https://en.cppreference.com/w/cpp/utility/from_chars#Notes :

> Unlike other parsing functions in C++ and C libraries, std::from_chars is locale-independent, non-allocating, and non-throwing.

laanwj commented on src/util/strencodings.h:
```diff
@@ -69,7 +71,25 @@ std::string EncodeBase32(const std::string& str, bool pad = true);
+{
+    T result;
+    // Emulate atoi(...) handling of white space and leading +/-.
+    std::string s = TrimString(str);
```
Making a copy of the string here, as well as below for the +/- check, is kind of efficient.
Not sure it matters here, it's not like this function is used in any performance critical places at the moment.

practicalswift replied:
As a general rule I try to stay away from pointer arithmetic ([rationale here](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es42-keep-use-of-pointers-simple-and-straightforward)), that's the reason for the technically extraneous copy which I think is insignificant in this context :)

Personally I find code that avoids pointer arithmetic _much_ easier to reason about from a safety perspective.

That said: I'll let other chime in :)

laanwj commented:
I think we need to make a choice here (either one or both):

- Make an emulation of all broken and surprising `atoi` behavior, but call it something else than `ToIntegral`, something like `LocaleIndependentAtoi`, make it clear it should not be used in new code.
- Make a sane integer parsing function that can eventually replace `Parse[U]IntXX` or their contents in a locale-independent way. I'm fine with the name then.

practicalswift commented:
@laanwj

> I think we need to make a choice here (either one or both):
> 
> * Make an emulation of all broken and surprising `atoi` behavior, but call it something else than `ToIntegral`, something like `LocaleIndependentAtoi`, make it clear it should not be used in new code, it's only meant for preserving `atoi` behavior for backward compatibility reasons.
> * Make a sane integer parsing function that can eventually replace `Parse[U]IntXX` or their contents in a locale-independent way. I'm fine with the name then.

Good point: I thought about those options as well. I think we should do both. This PR is meant as a pure refactoring: i.e. not changing any behaviour.

 I've now made it more clear that this PR is meant as a quirk-by-quirk compatible replacement for `atoi` by calling the function `LocaleIndependentAtoi` :)

Good point about naming: we should keep the name `ToIntegral` for a future sane version that doesn't emulate `atoi` or any other legacy function.

laanwj commented:
Okay, agree. I guess it could replace `atoi` as well as `atoi64` in that case (as it's parametrized on type)?

practicalswift commented:
@laanwj Done! I've now replaced all instances of `atoi64(s)` with `LocaleIndependentAtoi<int64_t>(s)`.

laanwj commented on src/util/strencodings.h:
```diff
@@ -68,8 +70,25 @@ std::string EncodeBase32(Span<const unsigned char> input, bool pad = true);
-int64_t atoi64(const std::string& str);
-int atoi(const std::string& str);
+
+template <typename T>
```
Please add a comment what this function does (e.g.: replicate the exact behavior of `atoi`) and that new code should use the `Parse[U]IntXX` functions with error checking.

practicalswift replied:
Done! :)

laanwj commented:
> @laanwj Done! I've now replaced all instances of atoi64(s) with LocaleIndependentAtoi<int64_t>(s).

Thanks, looks good to me now except the documentation nit.

laanwj commented:
I don't get the CI error. Is `charconv` non-standard in some way?
```
In file included from primitives/transaction.cpp:10:0:
./util/strencodings.h:16:10: fatal error: charconv: No such file or directory
 #include <charconv>
          ^~~~~~~~~~
compilation terminated.
```

practicalswift commented:
@laanwj I think the CI failures are from relatively older distros, and that C++17 library support lagged C++17 compiler support for a while in C++17 infancy. 

According to [cppreference](https://en.cppreference.com/w/cpp/compiler_support) the [elementary string conversions](https://wg21.link/P0067R5) (`<charconv>`) were added in GCC libstdc++ 8, Clang libc++ 7 and MSVC Standard Library 19.14.

MarcoFalke commented:
ok, this is blocked on #20460

practicalswift commented:
Could remove "Waiting for author"? I don't know if we have any "Blocked on another PR/issue" (or "Waiting for toolchain upgrade") tag, but this PR is blocked on #20460 :)

practicalswift commented:
Rebased!

Could this PR get a [23.0 milestone](https://github.com/bitcoin/bitcoin/milestone/52) like the related PR #20457 ("util: Make Parse{Int,UInt}{32,64} use locale independent std::from_chars(…) (C++17) instead of locale dependent strto{l,ll,ul,ull}")? :)

practicalswift commented:
Now that #20460 has been merged I think this PR should be ready for final review :)

See also related PR #20457.

practicalswift committed:
Replace use of locale dependent atoi(…) with locale-independent std::from_chars(…) (C++17)

test: Add test cases for LocaleIndependentAtoi

fuzz: Assert legacy atoi(s) == LocaleIndependentAtoi<int>(s)

fuzz: Assert legacy atoi64(s) == LocaleIndependentAtoi<int64_t>(s)

jonatack commented:
Concept ACK

laanwj commented:
Code review ACK 4343f114cc661cf031ec915538c11b9b030e2e15

jonatack commented on src/rpc/mining.cpp:
```diff
@@ -702,7 +702,7 @@ static RPCHelpMan getblocktemplate()
 
             hashWatchedChain = ParseHashV(lpstr.substr(0, 64), "longpollid");
-            nTransactionsUpdatedLastLP = atoi64(lpstr.substr(64));
+            nTransactionsUpdatedLastLP = LocaleIndependentAtoi<int64_t>(lpstr.substr(64));
```
`nTransactionsUpdatedLastLP` is `unsigned int`; should this be defined or cast to that?

laanwj replied:
Yes it probably needs the same kind of ugly cast as `nTimeSmart` in the wallet code. In the longer run it should imo be converted to an explicitly sized type, and parsed as that, but keeping the behavior exactly the same makes sense here to keep it uncontroversial.

practicalswift replied:
@jonatack Good point! I suggest we tackle that in a follow-up PR to keep this PR refactoring only :)

jonatack reviewed (commented):
Code review ACK 4343f114cc661cf031ec915538c11b9b030e2e15

laanwj commented on src/util/strencodings.h:
```diff
@@ -68,8 +69,33 @@ std::string EncodeBase32(Span<const unsigned char> input, bool pad = true);
+// LocaleIndependentAtoi is provided for backwards compatibility reasons.
+//
+// New code should use the ParseInt64/ParseUInt64/ParseInt32/ParseUInt32 functions
+// which provide parse error feedback.
```
This is a really important comment. I hope we can get rid of these functions entirely at some point. The implicit error handling behavior of `atoi` and `atoi64` is pretty much always undesirable.
(so I would normally comment "these function names are long and clunky, please shorten them" but it's good in this case, we want using them to be ugly :smile:  )

MarcoFalke commented on src/core_read.cpp:
```diff
@@ -69,7 +69,7 @@ CScript ParseScript(const std::string& s)
         {
             // Number
-            int64_t n = atoi64(*w);
+            int64_t n = LocaleIndependentAtoi<int64_t>(*w);
```
The condition literally checks for the required format for ParseIntegral<>, so why not use that?

practicalswift replied:
Sounds like a good idea now that `ParseIntegral` is part of `master` :)

MarcoFalke replied:
Sorry, I was wrong. This is not a refactor.

MarcoFalke commented on src/qt/rpcconsole.cpp:
```diff
@@ -250,7 +250,7 @@ bool RPCConsole::RPCParseCommandLine(interfaces::Node* node, std::string &strRes
                                         if (!IsDigit(argch))
                                             throw std::runtime_error("Invalid result query");
-                                    subelement = lastResult[atoi(curarg.c_str())];
+                                    subelement = lastResult[LocaleIndependentAtoi<int>(curarg)];
```
Same

MarcoFalke replied:
Sorry, I was wrong. This is not a refactor.

MarcoFalke commented on src/util/strencodings.h:
```diff
@@ -68,8 +69,33 @@ std::string EncodeBase32(Span<const unsigned char> input, bool pad = true);
+        }
+        s = s.substr(1);
+    }
+    auto [_, error_condition] = std::from_chars(s.data(), s.data() + s.size(), result);
```
why not `return ToIntegral<T>(s);`?

MarcoFalke commented on src/util/strencodings.h:
```diff
@@ -68,8 +69,33 @@ std::string EncodeBase32(Span<const unsigned char> input, bool pad = true);
+// LocaleIndependentAtoi is provided for backwards compatibility reasons.
+//
+// New code should use the ParseInt64/ParseUInt64/ParseInt32/ParseUInt32 functions
+// which provide parse error feedback.
```
The linter says to use `ToIntegral`, this contradicts by saying to use `ParseInt*`??

practicalswift replied:
`ToIntegral` sounds like a good suggestion now that it is part of `master` :)

MarcoFalke commented on src/util/moneystr.cpp:
```diff
@@ -77,8 +77,7 @@ std::optional<CAmount> ParseMoney(const std::string& money_string)
         return std::nullopt;
-    int64_t nWhole = atoi64(strWhole);
-
+    int64_t nWhole = LocaleIndependentAtoi<int64_t>(strWhole);
```
same

MarcoFalke replied:
Sorry, I was wrong. This is not a refactor.

maflcko reviewed (commented):
concept ack

MarcoFalke commented:
Follow up in #23184

maflcko reviewed (commented):
withdrawn my comments
